src/python/pants/backend/python/sandboxed_interpreter.py

```diff
@@ -0,0 +1,64 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+
+from pex.interpreter import PythonInterpreter
+
+from pants.util.memo import memoized_method
+
+
+class SandboxedInterpreter(PythonInterpreter):
+
+  class ToolchainLocationError(Exception):
+    def __init__(self, dir_path):
+      msg = "path '{}' does not exist or is not a directory".format(dir_path)
+      super(ToolchainLocationError, self).__init__(msg)
+
+  class BaseInterpreterError(Exception): pass
+
+  # using another PythonInterpreter to populate the superclass constructor args
+  def __init__(self, llvm_base_dir, base_interp):
+
+    if not os.path.isdir(llvm_base_dir):
+      raise ToolchainLocationError(llvm_base_dir)
+    if not isinstance(base_interp, PythonInterpreter):
+      raise BaseInterpreterError(
+        "invalid PythonInterpreter: '{}'".format(repr(base_interp)))
+
+    self._llvm_base_dir = llvm_base_dir
+
+    # this feels a little hacky -- what if pex's PythonInterpreter later needs
+    # another constructor arg that's not just a property of the class?
+    super(SandboxedInterpreter, self).__init__(
+      base_interp.binary, base_interp.identity, extras=base_interp.extras)
+
+  # made into an instance method here (unlike PythonInterpreter superclass) to
+  # use instance property self._llvm_base_dir
+  @memoized_method
+  def sanitized_environment(self):
+    sanitized_env = super(SandboxedInterpreter, self).sanitized_environment()
+
+    # use our compiler at the front of the path
+    # TODO: when we provide ld and stdlib headers, don't add the original path
+    sanitized_env['PATH'] = ':'.join([
+      os.path.join(self._llvm_base_dir, 'bin'),
+      os.environ.get('PATH'),
+    ])
+
+    # TODO: figure out whether we actually should be compiling fat binaries.
+    # this line tells distutils to only compile for 64-bit archs -- if not, it
+    # will attempt to build a fat binary for 32- and 64-bit archs, which makes
+    # clang invoke "lipo", an osx command which does not appear to be open
+    # source. see Lib/distutils/sysconfig.py and Lib/_osx_support.py in CPython.
+    sanitized_env['ARCHFLAGS'] = '-arch x86_64'
```

Comment:
Maybe I'm not understanding which things will run in this sandbox, but this seems specific to a particular usage of this class (for invoking `setup.py`). Would it be better for `build_local_python_distributions.py` to be able to set additional env vars?

Reply:
I actually had the contents of `sandboxed_interpreter.py` as a private contextmanager method of `BuildLocalPythonDistributions` and I think I thought that decoupling the "sandbox" environment from the task that directly consumes it was a useful goal, but it's not clear to me how useful that is now. It's easy enough to move back and thinking about it again, I don't know if I really like the idea of overriding an `@classmethod` with an instance method at all (which is what's happening with `sanitized_environment()`). I think it's probably better to put it back in the `BuildLocalPythonDistributions` task.

Reply:
Thinking about it again, it's likely because I wanted to take advantage of [pex's battle scars](https://github.com/pantsbuild/pex/blob/23709949e82a74fa02ad5629de47adf15a2c90fd/pex/interpreter.py#L329), and thought that subclassing the `PythonInterpreter` class in this way would do that cleanly. That being said, the `@classmethod` turning into an instance method is probably not a good idea at all (we're relying on it getting invoked as `self.sanitized_environment()` instead of `cls.sanitized_environment()`, unless we change the pex code to just be an instance method as well.

Reply:
One way of setting necessary env variables temporarily could be using `environment_as` from contextutil as I see you have imported below. I may be missing context here, but an alternative to PATH prepending could be to set `CC` and `CXX` (and any other env vars) using an `environment_as` wrapper around 
```
setup_runner = SetupPyRunner(dist_target_dir, 'bdist_wheel', interpreter=sandboxed_interpreter)
setup_runner.run()
```
Just a thought though, it seems like there are a few ways to go about this.

Reply:
I think that not baking the sandboxing into pex would be good... python/setup.py invokes will not be the only place we want to use it. So yea, `environment_as` (if possible) would be more general, and more portable to remote execution.

Reply:
Yep, that's what I was thinking -- a contextmanager private method would be applying a few levels of contexts from contextutil and then yielding. The reasons I liked doing that in the interpreter subclass are related to the utilities the pex interpreter class already provides, but if that's not significant there's no reason to do it that way.

Reply:
Ok, there's no hidden significance from my end behind tying this to the pex interpreter. I can push up a change today.